feat: Update client to Polkadot v1.7.2

wischli opened the pull request:
# Description

* Updates client to Polkadot v1.7.2
* Fixes open issues from v1.1.0 upgrade
* BREAKING: Replaces deprecated `export-genesis-state` with `export-genesis-head`
* Unifies `start_node` for all runtimes
* Switches ephemeral genesis chain spec generation to new `serde_json` approach
  * Enforces and enables us to drop all `Default::default` code paths in the genesis spec
* Feature creep: Updates external dependencies to latest supported state in Polkadot v1.7.2

## Tested

- [x] Create chain spec
- [x] Build blocks with local parachain client

# Checklist:

- [x] I have added Rust doc comments to structs, enums, traits and functions
- [x] I have made corresponding changes to the documentation
- [x] I have performed a self-review of my code
- [x] I have added tests that prove my fix is effective or that my feature works

wischli commented:
@lemunozm This compiles for all runtimes and builds blocks locally! Keeping the PR in draft mode until I have gone through it myself again and made some justifications about the changes. Sorry about the huge `Cargo.lock` diff, I upgraded all external dependencies to the versions supported in Polkadot v1.7.2.

wischli commented on node/src/chain_spec.rs:
```diff
@@ -635,88 +503,72 @@ fn development_genesis(
-		},
-		orml_tokens: development_runtime::OrmlTokensConfig {
-			balances: token_balances,
+	serde_json::json!({
```
This is new. The "old" way is still possible but this enables us to remove all the `Default::default()` code paths and thus also the amount of required boilerplate when adding new pallets with some genesis config.

lemunozm replied:
Yeah! That's great!

lemunozm commented:
Ping me when it's fully ready to be reviewed and I make a deep review. Thanks again for your help with the node 🥇

wischli commented on Cargo.toml:
```diff
@@ -54,33 +54,34 @@ repository = "https://github.com/centrifuge/centrifuge-chain"
-parity-scale-codec = { version = "3.2.2", default-features = false, features = ["derive"] }
-scale-info = { version = "2.3.0", default-features = false, features = ["derive"] }
-log = { version = "0.4", default-features = false }
+smallvec = "1.11.0"
+serde = { version = "1.0.195", default-features = false, features = ["derive"] }
+serde_json = { version = "1.0.111" }
+parity-scale-codec = { version = "3.6.1", default-features = false, features = ["derive"] }
+scale-info = { version = "2.10.0", default-features = false, features = ["derive"] }
+log = { version = "0.4.20", default-features = false }
```
Note: As mentioned in the PR description, I have bumped all of these to the versions used in Polkadot SDK v1.7.2

wischli commented on node/src/service.rs:
```diff
@@ -207,15 +199,10 @@ pub async fn start_altair_node(
 
-			let mut module = rpc::create_full(client.clone(), pool.clone(), deny_unsafe)?;
-			module
-				.merge(Anchors::new(client.clone()).into_rpc())
-				.map_err(|e| sc_service::Error::Application(e.into()))?;
```
NOTE: Moved to `create_full`

wischli commented on node/src/service.rs:
```diff
@@ -353,410 +267,105 @@ pub fn build_centrifuge_import_queue(
-		},
-	)
-	.await
+	let params = BasicAuraParams {
+		create_inherent_data_providers: move |_, ()| async move { Ok(()) },
+		block_import,
+		para_client: client,
+		relay_client: relay_chain_interface,
+		sync_oracle,
+		keystore,
+		collator_key,
+		para_id,
+		overseer_handle,
+		slot_duration,
+		relay_chain_slot_duration,
+		proposer,
+		collator_service,
+		// Very limited proposal time.
+		authoring_duration: Duration::from_millis(500),
+		collation_request_receiver: None,
+	};
```
The entire scope of this function including these parameters is taken from the parachain template in the Polkadot SDK: https://github.com/paritytech/polkadot-sdk/blob/0d7ca59fd4b4d15f9400971f32bc6a51b2fe837f/cumulus/parachain-template/node/src/service.rs#L400

lemunozm replied:
Thanks for the reference!

wischli commented on runtime/altair/src/lib.rs:
```diff
@@ -2694,6 +2696,16 @@ impl_runtime_apis! {
 		}
 	}
+
+	impl sp_genesis_builder::GenesisBuilder<Block> for Runtime {
```
NOTE: Required for the new genesis builder API

wischli commented on node/src/chain_spec.rs:
```diff
@@ -777,13 +616,13 @@ fn asset_registry_assets() -> Vec<(CurrencyId, Vec<u8>)> {
 						PalletInstance(parachains::rococo::rocksmine::usdt::PALLET_INSTANCE),
 						GeneralIndex(parachains::rococo::rocksmine::usdt::GENERAL_INDEX),
-					),
+					])),
```
Fixed: [96195c3](https://github.com/centrifuge/centrifuge-chain/pull/1844/commits/96195c36bb9428b95c230e8d6ed754ac6145609e)

wischli commented:
> The cargo lock lines added are just the ones I had removed in the base branch 😆, so everything is ok 👍🏻

I rebased and now there's "only" 2k new lines in the lock file 🫠

lemunozm commented:
> I rebased and now there's "only" 2k new lines in the lock file 🫠

Yeah, I uncommented integration-tests and all crates regarding this are now again in the repo

lemunozm reviewed (approved):
You did an amazing work on this @wischli!

Finally the service is much more legible and manageble

wischli commented on Cargo.lock:
```diff
@@ -654,6 +655,15 @@ version = "0.3.7"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "6b4930d2cb77ce62f89ee5d5289b4ac049559b1c45539271f5ed4fdc7db34545"
 
+[[package]]
+name = "arrayvec"
+version = "0.4.12"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "cd9fd44efafa8690358b7408d253adf110036b88f55672a933f01d616ad9b1b9"
+dependencies = [
+ "nodrop",
+]
+
```
I had some double dependencies after rebasing. I then checked out the `Cargo.lock` from the brase branch `polkadot-v1.7.2` and compiled my code.

wischli commented on node/src/service/evm.rs:
```diff
@@ -540,46 +534,13 @@ where
-		#[allow(deprecated)] // TODO fix before v1.3.0
-		start_full_node(params)?;
+		)?;
+	} else {
```
Great catch! Can be removed because it represents a fullnode which doesn't need to to consensus work. Base: https://github.com/paritytech/polkadot-sdk/blob/0d7ca59fd4b4d15f9400971f32bc6a51b2fe837f/cumulus/parachain-template/node/src/service.rs#L304

wischli commented on node/src/service/evm.rs:
```diff
@@ -527,10 +504,27 @@ where
+		para_id,
+		relay_chain_interface: relay_chain_interface.clone(),
+		task_manager: &mut task_manager,
+		da_recovery_profile: if validator {
+			DARecoveryProfile::Collator
+		} else {
+			DARecoveryProfile::FullNode
+		},
```
A fullnode is a superset of a collator node. The name comes from the fact that they do the full verification of all blocks. Moreover, they also hold storage of previous blocks in the database.

Collators maintain a “full-node” for a particular parachain; meaning they retain all necessary information to be able to author new blocks and execute transactions in much the same way as miners do on current PoW blockchains. Under normal circumstances, they will collate and execute transactions to create an unsealed block and provide it, together with a proof of state transition, to one or more validators responsible for proposing a parachain block.

The only non-collator fullnode I can think of are archive nodes (e.g. `wss://fullnode.centrige.io`) which users can interact with and serve as source of truth for indexers such as Subscan.

lemunozm replied:
Ah ok, thanks for the explanation. I thought the full node somehow was not a collator, but it is 👍🏻

wischli commented on node/src/service.rs:
```diff
@@ -48,6 +49,40 @@ type FullBackend = TFullBackend<Block>;
 type ParachainBlockImport<RuntimeApi, Executor> =
 	TParachainBlockImport<Block, Arc<FullClient<RuntimeApi, Executor>>, FullBackend>;
 
+pub trait RuntimeApiCollection:
+	sp_transaction_pool::runtime_api::TaggedTransactionQueue<Block>
+	+ sp_api::ApiExt<Block>
+	+ sp_block_builder::BlockBuilder<Block>
+	+ substrate_frame_rpc_system::AccountNonceApi<Block, AccountId, Nonce>
+	+ pallet_transaction_payment_rpc_runtime_api::TransactionPaymentApi<Block, Balance>
+	+ sp_api::Metadata<Block>
+	+ sp_offchain::OffchainWorkerApi<Block>
+	+ sp_session::SessionKeys<Block>
+	+ fp_rpc::ConvertTransactionRuntimeApi<Block>
+	+ fp_rpc::EthereumRuntimeRPCApi<Block>
+	+ sp_consensus_aura::AuraApi<Block, AuraId>
+	+ runtime_common::apis::AnchorApi<Block, Hash, BlockNumber>
+	+ cumulus_primitives_core::CollectCollationInfo<Block>
+{
+}
```
Credits go to the Moonbeam team! I love it too

wischli commented on node/src/chain_spec.rs:
```diff
@@ -146,30 +145,26 @@ pub fn centrifuge_local(para_id: ParaId) -> CentrifugeChainSpec {
+	.with_id("centrifuge_local")
+	.with_chain_type(ChainType::Local)
+	.with_genesis_config_patch(centrifuge_genesis(
+		vec![(
+			get_account_id_from_seed::<sr25519::Public>("Alice"),
+			get_from_seed::<AuraId>("Alice"),
+		)],
+		endowed_accounts(),
+		endowed_evm_accounts(),
+		Some(100000000 * CFG),
```
Yes, Alice's private key is well known. That's the point of it. All these ephemeral configs which use Alice for sensitive permissions are just local chains which have been super useful for testing runtime upgrades pre chopsticks. These days, they are used less but we should still keep them. Without them, you could not run a local Altair or Centrifuge Chain.

lemunozm replied:
But our production centrifuge-chain doesn't have any Alice, right?

wischli replied:
No, it does not!

lemunozm commented:
@wischli you can hit the `squash and merge` button whenever you think it's ready

mustermeiszer commented on node/src/rpc/evm.rs:
```diff
@@ -39,8 +39,24 @@ use sp_core::H256;
+	// NOTE: Not used by our precompiles right now. Therefore, no need to provide impl.
+	type EstimateGasAdapter = ();
+	// Assumes the use of HashedMapping<BlakeTwo256> for address mapping
+	type RuntimeStorageOverride =
```
What is this type doing?

wischli replied:
AFAIU, the eth config serves as an override adapter for `eth_call` similar to [geth](https://geth.ethereum.org/docs/interacting-with-geth/rpc/ns-eth). 

I have to admit that I am not confident in using the `SystemAccountId32StorageOverride` ([-> impl](https://github.com/moonbeam-foundation/frontier/blob/7f424fb9fc1ca9158a59be54cb183beaff9f799a/client/rpc/src/lib.rs#L140-L141)) type or whether `()` (i.e. no override) is the way to go. This was inspired [by Astar](https://github.com/AstarNetwork/Astar/blob/111d18fbfba6076570cb6454573f88afe0ee53b3/bin/collator/src/rpc.rs#L103).

WDYT @mustermeiszer?

Using the `ethereumRuntimeRPCApi` runtime API, I could not notice any difference between using `SystemAccountId32StorageOverride` and `()`. However, I did not execute any EVM call.

mustermeiszer reviewed (approved):
LGTM!
